File: src/de/dca/entitytags/util/ReflectionUtil.java
package de.dca.entitytags.util;

import sun.misc.Unsafe;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

public class ReflectionUtil {

    private static final Field fieldModifiersField;
    private static final Unsafe unsafeInstance;

    static {
        try{
            Field unsafeField = Unsafe.class.getDeclaredField("theUnsafe");
            unsafeField.setAccessible(true);
            unsafeInstance = (Unsafe) unsafeField.get(null);

            fieldModifiersField = Field.class.getDeclaredField("modifiers");
            fieldModifiersField.setAccessible(true);
        }catch (Throwable t){
            throw new RuntimeException(t); //Should never occur
        }
    }

    public static void setPrivateFinalField(Field f, Object object, Object value) throws IllegalAccessException {
        f.setAccessible(true);
        fieldModifiersField.setInt(f, f.getModifiers() & ~Modifier.FINAL);
        f.set(object, value);
    }

    public static <T> T allocateObject(Class<T> clazz) throws InstantiationException {
        return (T) unsafeInstance.allocateInstance(clazz);
    }

    public static <T> T copyObject(T src) throws InstantiationException, IllegalAccessException {
        return copyObject(src, src.getClass());
    }

    public static <T> T copyObject(T src, Class<?> typeToCopy) throws InstantiationException, IllegalAccessException {
        T out = (T) unsafeInstance.allocateInstance(typeToCopy);
        copyObjectInto(src, out);
        return out;
    }

    public static <T> void copyObjectInto(T from, T to) throws IllegalAccessException {
        copyObjectInto(from, to, from.getClass());
    }

    public static <T> void copyObjectInto(T from, T to, Class<?> typeToCopy) throws IllegalAccessException {
        for(Field f : typeToCopy.getDeclaredFields()){
            if((f.getModifiers() & Modifier.STATIC) == Modifier.STATIC){
                continue;
            }
            f.setAccessible(true);
            ReflectionUtil.setPrivateFinalField(f, to, f.get(from));
        }
    }
}

File: src/de/dca/entitytags/util/ReflectionUtil.kt
package de.dca.entitytags.util

import de.dca.entitytags.extensions.isStatic
import sun.misc.Unsafe
import java.lang.reflect.Field
import java.lang.reflect.Modifier

object ReflectionUtil {

    val unsafeInstance : Unsafe
    private val fieldModifiersField : Field

    init {
        unsafeInstance = findUnsafe()
        fieldModifiersField = findModifierField()
        fieldModifiersField.isAccessible = true
    }

    private fun findUnsafe() : Unsafe {
        try {
            val field = Unsafe::class.java.getDeclaredField("theUnsafe")
            field.isAccessible = true
            return (field.get(null) as Unsafe)
        }catch (ex: Exception){
            throw RuntimeException(ex)
        }
    }

    private fun findModifierField() : Field {
        try {
            return Field::class.java.getDeclaredField("modifiers")
        }catch (ex: Exception){
            throw RuntimeException(ex)
        }
    }

    fun setPrivateFinalField(field: Field, obj: Any, value: Any){
        field.isAccessible = true
        val modifiers = fieldModifiersField.getInt(field)
        fieldModifiersField.setInt(field, modifiers and Modifier.FINAL.inv())
        field.set(obj, value)
    }

    inline fun <reified T> allocateObject() : T {
        return (unsafeInstance.allocateInstance(T::class.java) as T)
    }

    fun <T : Any> copyObject(src: T) : T {
        return copyObject(src, src::class.java)
    }

    fun <T : Any> copyObject(src: T, typeToCopy: Class<*>) : T {
        val out = unsafeInstance.allocateInstance(typeToCopy) as T
        copyObjectInto(src, out)
        return out
    }

    fun <T : Any> copyObjectInto(src: T, dest: T){
        copyObjectInto(src, dest, src::class.java)
    }

    fun <T : Any> copyObjectInto(src: T, dest: T, typeToCopy: Class<*>) {
        for (field in typeToCopy.declaredFields){
            if (field.isStatic())
                continue
            field.isAccessible = true
            this.setPrivateFinalField(field, dest, field.get(src))
        }
    }
}